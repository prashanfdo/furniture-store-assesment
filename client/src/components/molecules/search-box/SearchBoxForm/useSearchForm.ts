import { useEffect } from "react";
import { useForm } from "react-hook-form";
import { useDebounce } from "usehooks-ts";

export type SearchFormFields = {
  search: string;
};

type useSearchFormProps = {
  initialSearchText: string;
  onSearchSubmit: (searchText: string) => void;
  onSearchChange: (searchText: string) => void;
};
const useSearchForm = ({ initialSearchText, onSearchSubmit, onSearchChange }: useSearchFormProps) => {
  const { register, handleSubmit, watch } = useForm<SearchFormFields>({
    defaultValues: {
      search: initialSearchText,
    },
  });

  const currentSearchText = watch("search");

  const debouncedCurrentSearchText = useDebounce<string>(currentSearchText, 300);

  useEffect(() => {
    onSearchChange(debouncedCurrentSearchText);
  }, [debouncedCurrentSearchText, onSearchChange]);

  const onSubmit = handleSubmit(({ search }) => onSearchSubmit(search));

  return { register, onSubmit };
};

export default useSearchForm;
