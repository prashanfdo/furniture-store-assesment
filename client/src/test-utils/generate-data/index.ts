export * from "./article";
export * from "./category";
export * from "./search";
