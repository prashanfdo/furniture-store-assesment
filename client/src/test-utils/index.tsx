import { render, RenderOptions, renderHook, RenderHookOptions, RenderHookResult } from "@testing-library/react";
import fs from "fs";
import React, { FC, ReactElement } from "react";

import user from "@testing-library/user-event";
import { ProvidersForTests } from "providers";

type GraphQlMocks = React.ComponentProps<typeof ProvidersForTests>["mocks"];

type WrapperProps = React.ComponentProps<typeof ProvidersForTests> & {
  children: React.ReactNode;
};
const Wrapper: FC<WrapperProps> = ({ children, mocks, initialRoute }: WrapperProps) => {
  return (
    <ProvidersForTests mocks={mocks} initialRoute={initialRoute}>
      {children}
    </ProvidersForTests>
  );
};
const wrapper =
  (mocks: GraphQlMocks, initialRoute?: string) =>
  ({ children }: { children: React.ReactNode }) =>
    (
      <Wrapper mocks={mocks} initialRoute={initialRoute}>
        {children}
      </Wrapper>
    );

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper"> & { gqlMocks: GraphQlMocks; initialRoute: string },
) => {
  const view = render(ui, { wrapper: wrapper(options?.gqlMocks, options?.initialRoute), ...options });
  const style = document.createElement("style");
  style.innerHTML = fs.readFileSync("src/index.out.css", "utf8");
  document.head.appendChild(style);

  return view;
};

export const customRenderHook = <T, R>(
  callback: (props: T) => R,
  options?: Omit<RenderHookOptions<T>, "wrapper"> & { gqlMocks?: GraphQlMocks; initialRoute?: string },
): RenderHookResult<R, T> => {
  return renderHook(callback, { ...options, wrapper: wrapper(options?.gqlMocks, options?.initialRoute) });
};

export { faker } from "@faker-js/faker";
export * from "@testing-library/react";
export * from "./generate-data";
export { customRender as render };
export { customRenderHook as renderHook };
export { user };
