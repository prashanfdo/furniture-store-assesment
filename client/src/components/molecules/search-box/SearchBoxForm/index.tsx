import React from "react";
import { SearchInput, SuggestionsList } from "components/atoms";
import useSearchForm from "./useSearchForm";

type SearchBoxFormProps = Parameters<typeof useSearchForm>[0] & React.ComponentProps<typeof SuggestionsList>;

const SearchBoxForm: React.FC<SearchBoxFormProps> = ({
  initialSearchText,
  onSearchSubmit,
  onSearchChange,
  suggestions,
  showSuggestions,
}: SearchBoxFormProps) => {
  const searchForm = useSearchForm({
    initialSearchText,
    onSearchSubmit,
    onSearchChange,
  });

  return (
    <>
      <form
        className="search-input-form w-full md:w-[16rem] lg:w-[32rem] relative grid grid-cols-[1fr_auto] border-gray-200 border rounded-lg group"
        onSubmit={searchForm.onSubmit}
      >
        <SearchInput {...searchForm.register("search")} />
        <SuggestionsList suggestions={suggestions} showSuggestions={showSuggestions} />
      </form>
    </>
  );
};

export default SearchBoxForm;
