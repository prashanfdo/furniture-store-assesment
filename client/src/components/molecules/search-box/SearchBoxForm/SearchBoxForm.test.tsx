import { createSearchSuggestionList, render, screen, user } from "test-utils";
import SearchBoxForm from "./index";

const dataSuggestions = createSearchSuggestionList(10);
jest.mock("usehooks-ts", () => ({
  useDebounce: (text: string) => text,
}));

describe("SearchBoxForm", () => {
  beforeEach(() => {
    user.setup();
  });

  it("should render", () => {
    const { container } = render(
      <SearchBoxForm
        initialSearchText="first search"
        onSearchChange={jest.fn()}
        onSearchSubmit={jest.fn()}
        showSuggestions={true}
        suggestions={[]}
      />,
    );
    expect(container).toMatchSnapshot();
    expect(screen.getByRole("textbox")).toHaveValue("first search");
  });

  it("should render suggestion list when passed", () => {
    render(
      <SearchBoxForm
        initialSearchText="first search"
        onSearchChange={jest.fn()}
        onSearchSubmit={jest.fn()}
        showSuggestions={true}
        suggestions={dataSuggestions}
      />,
    );
    const links = screen.queryAllByRole("link", {
      hidden: true,
    });
    expect(links).toHaveLength(dataSuggestions.length);
    expect(links.map((el) => el.textContent)).toEqual(expect.arrayContaining(dataSuggestions.map((el) => el.name)));
  });

  it("should not render suggestion list when showSuggestions=false", () => {
    render(
      <SearchBoxForm
        initialSearchText="first search"
        onSearchChange={jest.fn()}
        onSearchSubmit={jest.fn()}
        showSuggestions={false}
        suggestions={dataSuggestions}
      />,
    );
    expect(screen.queryByRole("link")).not.toBeInTheDocument();
  });

  it("should invoke callbacks onchange, onsubmit, setShowSuggestions", async () => {
    const handleSearchChange = jest.fn();
    const handleSearchSubmit = jest.fn();

    render(
      <SearchBoxForm
        initialSearchText=""
        onSearchChange={handleSearchChange}
        onSearchSubmit={handleSearchSubmit}
        showSuggestions={false}
        suggestions={[]}
      />,
    );

    const searchText = "second search";
    const input = screen.getByRole<HTMLInputElement>("textbox");

    expect(input).toHaveValue("");

    await user.type(input, searchText);
    expect(input).toHaveValue(searchText);
    expect(handleSearchChange).toHaveBeenCalled();
    expect(handleSearchChange).toHaveBeenLastCalledWith(searchText);

    await user.type(input, "{enter}");
    expect(handleSearchSubmit).toHaveBeenCalledTimes(1);
    expect(handleSearchSubmit).toHaveBeenLastCalledWith(searchText);
  });
});
