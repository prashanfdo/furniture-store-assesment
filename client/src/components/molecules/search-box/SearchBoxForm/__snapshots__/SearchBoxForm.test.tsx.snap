// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBoxForm should render 1`] = `
<div>
  <form
    class="search-input-form w-full md:w-[16rem] lg:w-[32rem] relative grid grid-cols-[1fr_auto] border-gray-200 border rounded-lg group"
  >
    <input
      autocomplete="off"
      class="w-full px-3 py-2 bg-gray-200 border-transparent border-none rounded-tl-lg rounded-bl-lg outline-transparent"
      name="search"
      type="text"
    />
    <button
      class="grid w-10 text-white transition-colors border-transparent rounded-tr-lg rounded-br-lg place-items-center bg-coral-500 hover:bg-coral-300"
    >
      <svg
        aria-hidden="true"
        class="w-5 h-5 font-bold"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M9 3.5a5.5 5.5 0 100 11 5.5 5.5 0 000-11zM2 9a7 7 0 1112.452 4.391l3.328 3.329a.75.75 0 11-1.06 1.06l-3.329-3.328A7 7 0 012 9z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </form>
</div>
`;
